[ERE-1824] New option for current user to Change their Email Address

ppettitau opened the pull request.

ppettitau committed:
[ERE-1824] New option for current user to Change their Email Address

ScriptSmith commented on rdrf/rdrf/templates/user/change_email_address.html:
```diff
@@ -0,0 +1,78 @@
+    {% csrf_token %}
+
+    <div class="mb-3 row">
+      <span class="col-sm-3 col-form-label">Current email:</span>
```
Needs translate template tags

ppettitau replied:
Done.

ScriptSmith commented on rdrf/rdrf/users/views.py:
```diff
@@ -0,0 +1,48 @@
+            messages.add_message(self.request,
+                                 messages.SUCCESS,
+                                 f'{_("Email address has been updated successfully for user")}: {user.get_full_name()}')
+            return redirect('email_address_change')
```
I think having a new EmailNotification to send on successful change would be useful to curators, and to notify the user themselves at their old email.

ppettitau replied:
Any tips on what to say in this email notification?

i.e. something like...

> You're receiving this email because your recent request to change your email address has been completed. This is a courtesy message to let you know that this email address is no longer active. You can request to change your email address at any time by using the "Change Email Address" available from your menu.

ScriptSmith replied:
Yeah that's good. I would just be clear that it's coming from 'the registry' so they don't think it's from their email provider.

ppettitau replied:
Done.

ScriptSmith commented on rdrf/rdrf/users/forms.py:
```diff
@@ -0,0 +1,65 @@
+        if self.user.patient:
+            duplicate_patients = Patient.objects.exclude(email__iexact=self.user.patient.email).filter(email__iexact=email)
+
+            if duplicate_patients:
```
I think this could cause issues with registries that intentionally have duplicate patients enrolled temporarily while they try to reconcile issues. 

In MND, it means that a patient won't be able to change their email while the curator hasn't cleared up any of their duplicates. 

eresearchqut/mnd#40

ScriptSmith replied:
As per discussion, this is exceedingly unlikely to occur

ScriptSmith commented on rdrf/rdrf/users/views.py:
```diff
@@ -0,0 +1,48 @@
+            messages.add_message(self.request,
+                                 messages.SUCCESS,
+                                 f'{_("Email address has been updated successfully for user")}: {user.get_full_name()}')
+            return redirect('email_address_change')
```
For self-service email change, I think there needs to be some sort of 'activation' process to confirm they have access to the new address.

https://security.stackexchange.com/questions/234060/what-is-the-suggested-best-practice-for-changing-a-users-email-address

ppettitau replied:
Done.

ppettitau committed:
Add missing translation tags

ppettitau committed:
Merge branch 'next_release' into ERE-1824_New_Change_Email_Address_Page

ppettitau committed:
[ERE-1824] Add activation workflow to an email change request

ppettitau committed:
Merge branch 'next_release' into ERE-1824_New_Change_Email_Address_Page

ppettitau committed:
Update test zips

ppettitau committed:
[ERE-1824] Add e2e test for change email request

ppettitau committed:
[ERE-1824] Fix issue with array as default parameter

ppettitau commented on rdrf/rdrf/services/io/notifications/email_notification.py:
```diff
@@ -258,6 +259,25 @@ def process_notification(reg_code=None, description=None, template_data={}):
+    notes = EmailNotification.objects.filter(registry__code=reg_code, description=description)
+
+    if notes:
+        return process_notification(reg_code, description, template_data, default_recipient)
```
Need to check, but I think this will also send an activation email to the user's current email address, which is not the desired behaviour.

ppettitau replied:
No, it's all good as long as the person configuring the email notification leaves the recipient field blank.

ppettitau committed:
Expose username from mixin

ppettitau committed:
Fix error from email preferences views when user is already authenticated (without requiring token)

ppettitau committed:
Fix linting

ppettitau committed:
Send notification to old email address

ppettitau commented:
@ScriptSmith , I think I have now addressed all previous requested changes if you would like to have another look when you have time. Not urgent, as we have other things in the pipeline that I would like to go through before I bring this in (since this is a bigger feature).

ScriptSmith commented on rdrf/rdrf/security/mixins.py:
```diff
@@ -13,14 +13,31 @@ def test_func(self):
+    is_valid_token = False
+
+    def dispatch(self, request, *args, **kwargs):
+        from rdrf.auth.signed_url.util import check_token
```
Is there a reason to import during each dispatch rather than when the view is first loaded?

ppettitau replied:
Nope. Fixed

ScriptSmith commented on rdrf/rdrf/security/mixins.py:
```diff
@@ -13,14 +13,31 @@ def test_func(self):
         return self.request.user.is_superuser or self.request.user.is_staff
 
 
-class ReportAccessMixin(UserPassesTestMixin):
```
Should also remove from `docs/security/catalog.rst`, line 43

ppettitau replied:
Removed from `catalog.rst` and also added the Token mixin to the catalog

ScriptSmith commented on rdrf/rdrf/services/io/notifications/email_notification.py:
```diff
@@ -258,6 +259,25 @@ def process_notification(reg_code=None, description=None, template_data={}):
+        logger.debug('Registry notification not found, using default template')
+        email_body = default_template.render(Context(template_data).flatten())
+
+        from django.core.mail import send_mail
```
Can move this import to the top of the file?

ppettitau replied:
Yes

ScriptSmith commented on rdrf/rdrf/templates/registration/email_reset_completed.html:
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+<p>
+  {% trans "Hello" %} {{user_full_name}},<br />
+  {% trans "You're receiving this email because your recent request to change your email address in the" %} {{ registry|title }} {% trans "registry has been completed" %}.<br />
```
For current registry `name`s we have:

- `Sample Registry`
- `MiNDAUS Registry`
- `angelman`
- `ARRK`

Which means the word registry appears in the sentence twice for MND and the sample registry. We could reconcile that or change the wording

ppettitau replied:
I've changed the wording from "registry" to "site", so:
* the Sample Registry site
* the MiNDAUS Registry site
* the angelman site
* the ARRK site

What do you think?

ppettitau replied:
I also removed the `|title` so that there are no unexpected changes to the registry name casing. Angelman might want to look at title casing theirs.

ppettitau committed:
PR Fixes

ScriptSmith commented on rdrf/rdrf/services/io/notifications/email_notification.py:
```diff
@@ -238,18 +238,19 @@ def append(self, key, obj):
 
 
-def process_given_notification(notification, template_data={}):
+def process_given_notification(notification, template_data={}, additional_recipients=None):
```
It's not currently a bug, and was already like this, but we should fix the mutable object default arguments for these functions. So use `None` rather than `{}` and `[]`, and then set the values in the function.

Worth doing so that something downstream from this function doesn't think it's fine to add keys / entries, and then accidentally leak data.

Should also make this a lint rule if not already, if we do ERE-2047

ppettitau replied:
+1 To make this a lint rule, that would have saved me getting caught out with this myself (and as I seem to keep going when it's been long enough to forget that I shouldn't do it :) )

Fixed.

ScriptSmith commented on rdrf/rdrf/templates/registration/email_reset_activation.html:
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+<p>
+  {% trans "Hello" %} {{user_full_name}},<br />
```
Should use `blocktrans` so that translators can re-arrange templated values

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/rdrf/templates/registration/email_reset_completed.html:
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+<p>
+  {% trans "Hello" %} {{user_full_name}},<br />
```
`blocktrans`

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/rdrf/templates/user/change_email_address.html:
```diff
@@ -0,0 +1,162 @@
+                               {% endif %}" id="id_current_request_status">{% if is_expired %}EXPIRED{% else %}{{ current_request.status|upper }}{% endif %}</span>
+            {% if current_request.status == EmailChangeRequestStatus.PENDING %}
+              <div class="form-text">
+                  {% trans "You are required to click the activation link in the email we have sent to " %}
```
`blocktrans`

ppettitau replied:
Fixed

ppettitau committed:
PR Fixes:
- remove mutable object default arguments for process_notification methods
- make use of blocktrans
